### PR TITLE
`pack`: include contents of directories in `files` field

### PR DESCRIPTION
**Pull request by josephfrazier**
**Summary**

This makes it so that you don't have to put '/**' after a directory in the
`files` field of package.json to ensure that the contents of the directory
will be published.

Fixes https://github.com/yarnpkg/yarn/issues/2498
Fixes https://github.com/yarnpkg/yarn/issues/2942
Fixes https://github.com/yarnpkg/yarn/issues/2851

Includes and closes https://github.com/yarnpkg/yarn/pull/3170

**Test plan**

I modified one of the pre-existing pack tests to also include and check
for a file inside a directory listed in package.json.

### CODE REVIEW

**Comment by TrySound** [comment]
Fixes https://github.com/yarnpkg/yarn/issues/2986
Fixes https://github.com/yarnpkg/yarn/pull/3166

**Comment by arcanis** [comment]
A [test](https://ci.appveyor.com/project/kittens/yarn/build/2113/job/7pxs5e44eq109adb#L3046) seems to be failing on Windows, can you check?

**Comment by arcanis** [comment]
The Windows test still breaks.

Btw, do you support this edge case: `{ "files": [ "test/*" ] }`? Since it doesn't evaluate to a valid stat call it will get added as-is, so it probably won't pack any file beneath.

Maybe a better way to handle this would be to transform each `[ "file-name" ]` into `[ "file-name", "file-name/**" ]`, whether it's a file or a folder. Glob will filter them accordingly anyway.

**Comment by josephfrazier** [comment]
Yeah.... I don't have a Windows machine readily available to develop on, so I've been relying on AppVeyor as a bit of debugging tool (hence https://github.com/yarnpkg/yarn/pull/3175/commits/517d7df3cbd0642931eda84ba56f5eded4752eb6) :P

I just pushed a commit that tries the `[ "file-name" ]` into `[ "file-name", "file-name/**" ]` approach, and it looks like that gets a little closer to passing CI (only [one missing file](https://ci.appveyor.com/project/kittens/yarn/build/2137/job/ayfe6ncn81qq84us#L5135), rather than [two](https://ci.appveyor.com/project/kittens/yarn/build/2130/job/4hjfo6ke6dcsh2rw#L5160)). Thanks for the suggestion!

As for the `{ "files": [ "test/*" ] }` edge case, I tried changing the `dir/` into `dir/*` [here](https://github.com/yarnpkg/yarn/pull/3175/commits/6626c5e37ad66eec58143bf5bb481812258efbf2#diff-76aeddc6ab4277c9c808f7bdbe1684eeR6), and the tests still pass locally, so hopefully that won't be an issue, especially after switching to your recommended way of handling this.

I'm going to be mostly unavailable until next week, but I should be able to take another look at this then. In the mean time, feel free to push more commits onto this branch if you have some insight on how to fix the Windows tests.

**Comment by bestander** [comment]
Thanks a lot for sending the fix, @josephfrazier.
Looking forward to see the Windows tests pass.

**Comment by EnoahNetzach** [comment]
The main problem with Windows is the path separator: in places like [this one](https://github.com/josephfrazier/yarn/blob/cbdfee66af0b4f16d999af31eebb66a486a463e7/src/util/filter.js#L96) it is assumed it's `/` (in the regexs).

File: src/util/filter.js [hunk 1]
```diff
@@ -99,6 +99,7 @@ export function matchesFilter(filter: IgnoreFilter, basename: string, loc: strin
   }
   return filter.regex.test(loc) ||
          filter.regex.test(`/${loc}`) ||
+         filter.regex.test(`\\${loc}`) ||
```
Comment (EnoahNetzach) [inline src/util/filter.js]: why not using `path.sep`?

Comment (EnoahNetzach) [inline src/util/filter.js]: The problem I see with keeping both `/` and `\\` is that the backslash has a meaning in unix systems, so false-positives could occur.

Comment (josephfrazier) [inline src/util/filter.js]: The argument to `test` is a string rather than a RegExp pattern, so it shouldn't be interpreting backslashes as the escape character in this context. Note that the following expression (on Windows) should produce a string whose first character is a backslash:

```js
`${path.sep}${loc}`
```

That said, I'm not sure why https://github.com/yarnpkg/yarn/pull/3175/commits/c2df043343092dcc408f8792ad16eb86cad6ba3a seemed to make things worse, test-wise (an additional test [failed](https://ci.appveyor.com/project/kittens/yarn/build/2195/job/q5u26f85qlroy533#L3011))

**Comment by EnoahNetzach** [comment]
Printing all the regexs that are tested, I see a lot of unix separators:

```regex
(1) /^(?:\/(?=.)package\.json)$/i
(2) /^(?:\/(?=.)readme[^\/]*?)$/i
(3) /^(?:\/(?!\.)(?=.)(?:license|licence)+[^\/]*?)$/i 
(4) /^(?:\/(?!\.)(?=.)(?:changes|changelog|history)+[^\/]*?)$/i
(5) /^(?:(?=.)index\.js)$/i
(6) /^(?:(?=.)a\.js)$/i
(7) /^(?:(?=.)b\.js)$/i
(8) /^(?:(?=.)dir\/)$/i
(9) /^(?:(?=.)index\.js\/(?:(?!(?:\/|^)\.).)*?)$/i
(10) /^(?:(?=.)a\.js\/(?:(?!(?:\/|^)\.).)*?)$/i
(11) /^(?:(?=.)b\.js\/(?:(?!(?:\/|^)\.).)*?)$/i
(12) /^(?:(?=.)dir\/(?:(?!(?:\/|^)\.).)*?)$/i
(13) /^(?:(?!\.)(?=.)[^\/]*?)$/i
```

2, 3, 4, 8, 9, 10, 11, 12, and 13 all have a `\/`, whereas probably there should be a `\${path.sep}`, again..

Or am I overcomplicating things?

**Comment by josephfrazier** [comment]
The regexes are being generated in [ignoreLinesToRegex](https://github.com/josephfrazier/yarn/blob/c2df043343092dcc408f8792ad16eb86cad6ba3a/src/util/filter.js#L105), which [uses minimatch](https://github.com/josephfrazier/yarn/blob/c2df043343092dcc408f8792ad16eb86cad6ba3a/src/util/filter.js#L127) to build the regex, which normalizes paths to POSIX-style:

* https://github.com/isaacs/minimatch/blob/eed89491bd4a4e6bc463aac0dfb5c29ef0d1dc13/minimatch.js#L122-L125
* https://github.com/isaacs/minimatch/blob/eed89491bd4a4e6bc463aac0dfb5c29ef0d1dc13/minimatch.js#L711-L714

Accordingly, maybe we should change the logic in [matchesFilter](https://github.com/josephfrazier/yarn/blob/c2df043343092dcc408f8792ad16eb86cad6ba3a/src/util/filter.js#L96) to use [Minimatch#match](https://github.com/isaacs/minimatch/blob/eed89491bd4a4e6bc463aac0dfb5c29ef0d1dc13/minimatch.js#L699) instead of manually operating on the regexes.

**Comment by josephfrazier** [comment]
@bestander, @arcanis: The (relevant, non-flaky) tests are passing on all platforms now. Can you have another look when you get a chance? For reference, here are the changes since last week: https://github.com/yarnpkg/yarn/pull/3175/files/cbdfee66af0b4f16d999af31eebb66a486a463e7..37fc30aaf6f006a0a5f7a6fb731cca20c3d197a8

**Comment by arcanis** [comment]
Merged! Thanks! :)

**Comment by josephfrazier** [comment]
Thanks! I think this also fixes https://github.com/yarnpkg/yarn/pull/3166 (thanks for [noticing](https://github.com/yarnpkg/yarn/pull/3175#issuecomment-294710697), @TrySound!)

**Comment by TrySound** [comment]
@josephfrazier Thanks for PR :)
